File: python/dependency_injection.py
# Dependency injection: Don't instantiate other classes inside your class.
# Common approaches: constructor injection, setter injection, and method injection.

class Service:
    def serve(self):
        return "Service is running"


class MethodInjection:
    def run(self, service: Service):
        return service.serve()

class ConstructorInjection:
    def __init__(self, service: Service):
        self.service = service

    def run(self):
        return self.service.serve()

class SetterInjection:
    def __init__(self):
        self.service = None

    def set_service(self, service: Service):
        self.service = service

    def run(self):
        return self.service.serve()


if __name__ == "__main__":
    service = Service()

    client1 = MethodInjection()
    print("Method Injection:", client1.run(service))

    client2 = ConstructorInjection(service)
    print("Constructor Injection:", client2.run())

    client3 = SetterInjection()
    client3.set_service(service)
    print("Setter Injection:", client3.run())

File: c++/dependency_injection/dependency_injection.cpp
#include <iostream>
#include <string>

class Service {
public:
    std::string serve() {
        return "Service is running";
    }
};

class MethodInjection {
public:
    std::string run(Service* service) {
        return service->serve();
    }
};

class ConstructorInjection {
private:
    Service* service;
public:
    ConstructorInjection(Service* service) : service(service) {}

    std::string run() {
        return service->serve();
    }
};

class SetterInjection {
private:
    Service* service;
public:
    SetterInjection() : service(NULL) {}

    void setService(Service* service) {
        this->service = service;
    }

    std::string run() {
        return service->serve();
    }
};

int main() {
    std::cout << "--------------------------" << std::endl;
    std::cout << "Dependency injection" << std::endl;
    std::cout << "==========================" << std::endl;

    Service service;

    MethodInjection client1;
    std::cout << "Method Injection: " << client1.run(&service) << std::endl;

    ConstructorInjection client2(&service);
    std::cout << "Constructor Injection: " << client2.run() << std::endl;

    SetterInjection client3;
    client3.setService(&service);
    std::cout << "Setter Injection: " << client3.run() << std::endl;

    std::cout << "--------------------------" << std::endl << std::endl;
    return 0;
}
